vowpalwabbit/learner.cc

```diff
@@ -36,207 +30,251 @@ const char* to_string(prediction_type_t prediction_type)
+  bool ec_is_example_header(example& ec)
+  {
+    v_array<CB::cb_class> costs = ec.l.cb.costs;
+    if (costs.size() != 1)
+      return false;
+    if (costs[0].probability == -1.f)
+      return true;
     return false;
-  if (costs[0].probability == -1.f)
-    return true;
-  return false;
-}
-
-/* is this just a newline */
-inline bool example_is_newline_not_header(example& ec) { return (example_is_newline(ec) && !ec_is_example_header(ec)); }
-
-/* Adds an example to multiline collection
- * Returns: true if complete and false if incomplete example */
-bool complete_multi_ex(example* ec, multi_ex& ec_seq, vw& all)
-{
-  const bool is_test_ec = all.p->lp.test_label(&ec->l);
-
-  if (example_is_newline_not_header(*ec) && is_test_ec)
+  }
+  
+  bool inline is_save_cmd(example* ec)
   {
-    VW::finish_example(all, *ec);
-    if (ec_seq.size() == 0)
+    return (ec->tag.size() >= 4) && (0 == strncmp((const char*)ec->tag.begin(), "save", 4));
+  }
+  
+  /* is this just a newline */
+  inline bool example_is_newline_not_header(example& ec) { return (example_is_newline(ec) && !ec_is_example_header(ec)); }
+
+  void drain_examples(vw& all) {
+    if (all.early_terminate) { // drain any extra examples from parser.
+      example* ec = nullptr;
+      while ((ec = VW::get_example(all.p)) != nullptr) VW::finish_example(all, *ec);
+    }
+    all.l->end_examples();
+  }
+  
+  class single_instance_context {
+  public:
+    single_instance_context(vw& all)
+    : _all(all) {}
+    
+    vw& get_master() const {return _all;}
+    
+    template <class T, void (*process_impl)(T&, vw&)>
+    void process(T& ec) {
+      process_impl(ec, _all);
+    }
+  private:
+    vw& _all;
+  };
+  
+  class multi_instance_context {
+  public:
+    multi_instance_context(const std::vector<vw*>& all)
+    : _all(all) {}
+    
+    vw& get_master() const {return *_all.front();}
+    
+    template <class T, void (*process_impl)(T&, vw&)>
+    void process(T& ec) {
+      // start with last as the first instance will free the example as it is the owner
+      for (auto it = _all.rbegin(); it != _all.rend(); ++it) process_impl(ec, **it);
+    }
+  private:
+    std::vector<vw*> _all;
+  };
+  
+  template<typename context_type>
+  class single_example_handler {
+  public:
+    single_example_handler(const context_type& context)
+    : _context(context) {}
+    
+    void on_example(example* ec) {
+      if (ec->indices.size() > 1)  // 1+ nonconstant feature. (most common case first)
+        _context.template process<example, learn_ex>(*ec);
+      else if (ec->end_pass)
+        _context.template process<example, end_pass>(*ec);
+      else if (is_save_cmd(ec))
+        _context.template process<example, save>(*ec);
+      else
+        _context.template process<example, learn_ex>(*ec);
+    }
+  private:
+    context_type _context;
+  };
+  
+  template<typename context_type>
+  class multi_example_handler {
+  private:
+    bool complete_multi_ex(example* ec)
```

Comment:
I think I am misunderstanding but I don't see where multi_ex actually get's passed to learn/predict?